In [ ]:
!pip install tensorflow==2.1.0

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
ERROR: Could not find a version that satisfies the requirement tensorflow==2.1.0 (from versions: 2.5.0, 2.5.1, 2.5.2, 2.5.3, 2.6.0rc0, 2.6.0rc1, 2.6.0rc2, 2.6.0, 2.6.1, 2.6.2, 2.6.3, 2.6.4, 2.6.5, 2.7.0rc0, 2.7.0rc1, 2.7.0, 2.7.1, 2.7.2, 2.7.3, 2.7.4, 2.8.0rc0, 2.8.0rc1, 2.8.0, 2.8.1, 2.8.2, 2.8.3, 2.8.4, 2.9.0rc0, 2.9.0rc1, 2.9.0rc2, 2.9.0, 2.9.1, 2.9.2, 2.9.3, 2.10.0rc0, 2.10.0rc1, 2.10.0rc2, 2.10.0rc3, 2.10.0, 2.10.1, 2.11.0rc0, 2.11.0rc1, 2.11.0rc2, 2.11.0, 2.12.0rc0, 2.12.0rc1)
ERROR: No matching distribution found for tensorflow==2.1.0


In [ ]:
!pip install tensorflow

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [14]:
import tensorflow as tf
import numpy as np
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow import keras
print(tf.__version__)

2.11.0


In [ ]:
# 머신러닝 시작하기
dense = Dense(units=1, input_shape=[1])
model = Sequential([dense])
model.compile(optimizer='sgd', loss='mean_squared_error')

xs = np.array([-1.0, 0.0, 1.0, 2.0, 3.0, 4.0], dtype=float)
ys = np.array([-3.0, -1.0, 1.0, 3.0, 5.0, 7.0], dtype=float)

model.fit(xs, ys, epochs=500)

print(model.predict([10.0]))

텐서플로는 Sequential 클래스로 층을 정의한다.
keras.layers API를 사용하여 층을 정의할 수 있다.
컴퓨터는 입력할 데이터 간의 관계를 모르기 때문에 임의의 값으로 놓고 추측한 다음 손실함수에 따라 좋고 나쁨을 측정한다
추측에 사용되는 것을 옵티마이저라고 하며 위의 경우는 sgd라는 확률적 경사 하강법을 사용
추측한 값에서 오차를 계산할 결과가 주어졌을 때 또 다른 추측을 만드는 수학함수
과정 반복을 통해 손실을 최소화하며 추측한 값이 정답에 더 가깝게 만든다.
마지막으로는 지정한 epochs 만큼 학습을 반복한다.

모델이 10을 추측한 값이 19가 아닌 이유는
첫째, 손실이 0이 아니라 매우 작은 값이기에 그만큼 벗어날 것이라고 예상
둘째, 신경망이 학습한 데이터의 양이 매우 작음

In [ ]:
# 신경망이 학습한 것 확인하기

print('신경망이 학습한 것:{}'.format(dense.get_weights()))

신경망이 학습한 것:[array([[1.9975243]], dtype=float32), array([-0.99232435], dtype=float32)]


신경망이 학습한 관계는 Y = 1.9975243X - 0.99232435
기대했던 Y = 2X - 1과 매우 비슷하다는 것을 확인할 수 있다.

In [15]:
# 컴퓨터 비전 소개
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
])

활성화 함수란 층의 각 뉴런에 적용되는 함수

In [16]:
data = tf.keras.datasets.fashion_mnist

(training_images, training_labels), (test_images, test_labels) = data.load_data()

trainig_images = training_images / 255.0
test_images = test_images / 255.0

model = tf.keras.models.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(training_images, training_labels, epochs=5)

4422102/4422102 [==============================] - 0s 0us/step
Epoch 1/5
1875/1875 [==============================] - 6s 3ms/step - loss: 2.9761 - accuracy: 0.6921
Epoch 2/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.6339 - accuracy: 0.7706
Epoch 3/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5669 - accuracy: 0.8025
Epoch 4/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5207 - accuracy: 0.8220
Epoch 5/5
1875/1875 [==============================] - 6s 3ms/step - loss: 0.5018 - accuracy: 0.8285


정규화, 손실함수 희소한 범주형 크로스 엔트로피, 옵티마이저 adam,


In [17]:
model.evaluate(test_images, test_labels)

313/313 [==============================] - 1s 2ms/step - loss: 2.4164 - accuracy: 0.1000


[2.416425943374634, 0.10000000149011612]

In [18]:
classifications = model.predict(test_images)
print(classifications[0])
print(test_labels[0])

313/313 [==============================] - 1s 2ms/step
[0.1257129  0.05488903 0.23484713 0.08052519 0.04538276 0.08300455
 0.19524188 0.04969458 0.10679933 0.02390259]
9


In [24]:
class myCallback(tf.keras.callbacks.Callback):
  def on_epoch_end(self, epoch, logs={}):
    if(logs.get('accuracy')>0.95):
      print('\n정확도 95%에 도달하여 훈련을 멈춥니다!')
      self.model.stop_training = True

callbacks = myCallback()
mnist = tf.keras.datasets.fashion_mnist

(training_images, training_labels), (test_images, test_labels) = mnist.load_data()

training_images = training_images / 255.0
test_images = test_images / 255.0

model = tf.keras.models.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation=tf.nn.relu),
    keras.layers.Dense(10, activation=tf.nn.softmax)
])

model.compile(optimizer='adam',
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

model.fit(training_images, training_labels, epochs=50, callbacks=[callbacks])

Epoch 1/50
1875/1875 [==============================] - 7s 3ms/step - loss: 0.5026 - accuracy: 0.8232
Epoch 2/50
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3765 - accuracy: 0.8640
Epoch 3/50
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3374 - accuracy: 0.8768
Epoch 4/50
1875/1875 [==============================] - 6s 3ms/step - loss: 0.3137 - accuracy: 0.8861
Epoch 5/50
1875/1875 [==============================] - 7s 4ms/step - loss: 0.2966 - accuracy: 0.8900
Epoch 6/50
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2798 - accuracy: 0.8962
Epoch 7/50
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2697 - accuracy: 0.8992
Epoch 8/50
1875/1875 [==============================] - 6s 3ms/step - loss: 0.2583 - accuracy: 0.9040
Epoch 9/50
1875/1875 [==============================] - 5s 3ms/step - loss: 0.2483 - accuracy: 0.9071
Epoch 10/50
1875/1875 [==============================] - 6s 3ms/step - loss: 0.239